lib/libc/common/source/stdlib/malloc.c

```diff
@@ -153,6 +153,14 @@ void *aligned_alloc(size_t alignment, size_t size)
 	return ret;
 }
 
+#ifdef CONFIG_CPP
+/* G++ uses this function internally */
+void *memalign(size_t alignment, size_t size)
+{
+	return aligned_alloc(alignment, size);
```

Comment:
Probably should check if alignment is non-zero and is a power of two.

```
size > 0 && (size & (size - 1)) == 0
```

This might be checked already in zephyr's `k_aligned_alloc()`.

Also, since C is pretty high traffic, it would be great to add a couple of test cases.

One additional nit: please correct the commit message to ensure it says `aligned_alloc()` is part of ISO C. Maybe there was a mixup with the `posix_memalign()` function which is similar.

Reply:
Good points -- it's not technically valid to implement `memalign` as a direct call to `aligned_alloc` as `aligned_alloc `requires size to be a multiple of alignment while memalign does not. However, the manual for `k_heap_aligned_alloc` reads exactly like the manual for `memalign`, so it is valid to implement both of them with the same underlying code. It would be semantically cleaner to implement `memalign` in terms of `k_heap_aligned_alloc` and then `aligned_alloc` in terms of `memalign`, but as we want to expose only `aligned_alloc` in the Zephyr API, and `memalign` should only be used by the C++ code, the layering chosen here is slightly more efficient. for the desired common case.

If we used a symbol alias instead of a nested function call, then `memalign` would not introduce any additional overhead, but I couldn't find any instance in the Zephyr code using that mechanism.

I can add some comments in the code to make this all a bit more clear.

Reply:
I've copied the failing code from #57899 into a new test run as part of the existing libcxx tests. That allowed me to do some before/after testing to make sure this patch actually fixes the bug. And I've added more comments in the code to explain why the memalign call is allowed to just directly call aligned_alloc.